src/transformers/modeling_utils.py

```diff
@@ -2715,11 +2726,15 @@ def _fix_key(key):
 
         # retrieve unintialized modules and initialize before maybe overriding that with the pretrained weights.
         if _fast_init:
-            uninitialized_modules = model.retrieve_modules_from_names(
-                missing_keys, add_prefix=add_prefix_to_model, remove_prefix=remove_prefix_from_model
-            )
-            for module in uninitialized_modules:
-                model._init_weights(module)
+            if remove_prefix_from_model:
+                _loaded_keys = [f"{prefix}.{k}" for k in loaded_keys]
+            elif add_prefix_to_model:
+                _loaded_keys = [k[len(prefix) + 1 :] for k in loaded_keys]
+            else:
+                _loaded_keys = loaded_keys
+            set_initialized_submodules(model, _loaded_keys)
+            # This will only initialize submodules that are not marked as initialized by the line above.
+            model.apply(model._init_weights)
```

Comment:
would this work w/o needing to do any changes to the actual models?
```
def wrapper(module):
  if getattr(module, "_is_hf_initialized", False):
    return
  model._init_weights(module)
  module._is_hf_initialized = True

model.apply(wrapper)
```

Reply:
the problem with changing the models is that users who have already written the modeling code won't know to add this extra code.

Reply:
or replacing `apply` with an actual traversal of the submodule-tree which wold be the same as the wrapper suggestion, but even easier to read.